core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java

```diff
@@ -47,6 +51,70 @@
   // the number of seconds before we allow a thread to terminate with non-use.
   public static final long DEFAULT_TIMEOUT_MILLISECS = 180000L;
 
+  private static final ThreadPoolExecutor SCHEDULED_FUTURE_CHECKER_POOL =
+      createFixedThreadPool(1, "Scheduled Future Checker", false);
+
+  private static final ConcurrentLinkedQueue<ScheduledFuture<?>> CRITICAL_RUNNING_TASKS =
+      new ConcurrentLinkedQueue<>();
+
+  private static Runnable CRITICAL_TASK_CHECKER = new Runnable() {
+    @Override
+    public void run() {
+      while (true) {
+        Iterator<ScheduledFuture<?>> futures = CRITICAL_RUNNING_TASKS.iterator();
+        while (futures.hasNext()) {
+          ScheduledFuture<?> criticalTask = futures.next();
+          // Calling get() on a ScheduledFuture will block unless that scheduled task has
+          // completed. We call isDone() here instead. If the scheduled task is done then
+          // either it was a one-shot task, cancelled or an exception was thrown.
+          if (criticalTask.isDone()) {
+            // Now call get() to see if we get an exception.
+            try {
+              criticalTask.get();
+              // If we get here, then a scheduled task exited but did not throw an error
+              // or get canceled. This was likely a one-shot scheduled task.
+              if (!CRITICAL_RUNNING_TASKS.remove(criticalTask)) {
+                LOG.warn("Unable to remove task from list of watched critical tasks");
+              }
+            } catch (ExecutionException ee) {
+              // An exception was thrown in the critical task. Throw the error here, which
+              // will then be caught by the AccumuloUncaughtExceptionHandler which will
+              // log the error and terminate the VM.
+              throw new RuntimeException("Critical scheduled task failed.", ee.getCause());
```

Comment:
Calling ee.getCause() drops the stack trace associated with the ExecutionException which may contain important info for debugging problems.   The uncaughtexceptionhandler only halts the VM when its an error.  Since this is not an error then them VM will not be halted and this thread that checks future will be dead. 

```suggestion
              throw new RuntimeException("Critical scheduled task failed.", ee);
```

Reply:
> The uncaughtexceptionhandler only halts the VM when its an error. Since this is not an error then them VM will not be halted and this thread that checks future will be dead.

Good point - I forgot about that. Since this is deemed a critical background thread, do you think we should throw an Error here to terminate the VM?

Reply:
Thinking of throwing com.google.common.util.concurrent.ExecutionError here...

Reply:
See ccd2673

Reply:
Wondering if we should do something like the following in UEH to cover the case where a cause is an Error.

```java
class AccumuloUncaughtExceptionHandler implements UncaughtExceptionHandler {

  private static final Logger LOG = LoggerFactory.getLogger(AccumuloUncaughtExceptionHandler.class);

  private static boolean isError(Throwable t) {
    while(t != null) {
      if (t instanceof Error) {
        return true;
      }
      
      t = t.getCause();
    }
    
    return false;
  }
  
  @Override
  public void uncaughtException(Thread t, Throwable e) {
    if (e instanceof Exception) {
      LOG.error("Caught an Exception in {}. Thread is dead.", t, e);
    } else if isError(e) {
      try {
        e.printStackTrace();
        System.err.println("Error thrown in thread: " + t + ", halting VM.");
      } catch (Throwable e1) {
        // If e == OutOfMemoryError, then it's probably that another Error might be
        // thrown when trying to print to System.err.
      } finally {
        Runtime.getRuntime().halt(-1);
      }
    }
  }
}

```

Reply:
unresolved comment because I made a comment unrelated to the original problem... but the comment was inspired by the problem

Reply:
My thinking is that if these are critical background threads, then regardless of whether it's an Exception or an Error, the critical background task that should be running, isn't. I didn't wrap all scheduled tasks with this method, just the critical ones.

Reply:
This code looks good.  I was just thinking about the case where we have something like `new RuntimeException(<some error>)` happenning somewhere else in the code and making it to the UEH, would be nice if it unraveled it and found the error.